In [1]:
# Permet de tout executer au lancement du notebook + conserver le notebook actif pendant 2h
from IPython.display import Javascript
from masquer import *
Javascript("""
function repeter(){
IPython.notebook.kernel.execute("a=1");
}
// execute a = 1 en python toutes les 8 minutes pendant 2h
let timerId = setInterval(() => repeter(), 4800);
setTimeout(() => { clearInterval(timerId); alert('fin de cession'); }, 7200000);

// Supprimer la taille limite pour la sortie d'une cellule
IPython.OutputArea.prototype._should_scroll = function(lines) {
    return false;
};
IPython.notebook.kernel.execute("url = '" + window.location + "'");

// Exécuter toutes les cellule du notebook
    require(
        ['base/js/namespace', 'jquery'], 
        function(jupyter, $) {
            
                
                jupyter.actions.call('jupyter-notebook:run-all-cells-below');
                jupyter.actions.call('jupyter-notebook:save-notebook');
                Jupyter.actions.call('jupyter-notebook:hide-header')

        }
    );""")

<IPython.core.display.Javascript object>

# <span style="color:red;"><center> TD - Les changements de base </span>

## 1. Conversion d'une base b vers la base décimale

### 1.1 Méthode de droite à gauche

Voici un algorihtme permettant de convertir un nombre entier écrit en base $b$ vers la base décimale :
> resultat $\leftarrow$ 0<br>
> i $\leftarrow$ 0<br>
> **Pour** tous les chiffres $c$ allant de droite à gauche : <br>
>>    resultat $\leftarrow$ resultat + $c \times b^i$<br>
>>    i $\leftarrow$ i+1<br>
>
> Renvoyer resultat<br>

**Q1. Programmer une fonction ``convert_dec`` qui, selon l'algorithme précédent, prend en argument un nombre entier et la base dans laquelle il est écrit et renvoie sa valeur décimale. Penser à spécifier la fonction et éventuellement à utiliser ``assert``.**

In [46]:
# Question 1
c=(input("Entrer un nombre dans la base d'origine: "))
b=int(input("Donner la base de votre nombre: "))
    
def convert_dec(c,b):
    """Cette fonction permet de convertir un nombre d'une base spécifiée à la base décimale (pas hexadecimal)."""
    assert b>=2, "Base doit être supérieure ou égale à 2"
    resultat=0
    i=0
    l=len(c)
    
    for i in range (l):
        resultat=resultat+int(c[l-1-i])*b**i
        #print (resultat, i)
    return resultat
    
result=convert_dec(c,b)
print(result)

Entrer un nombre dans la base d'origine: 10101
Donner la base de votre nombre: 2
21


**Q2. Utiliser la fonction``convert_dec`` précédente pour donner la valeur décimale des nombres suivants :**
$101001_2$, $110100_2$, $123_4$, $1230_4$, $1010_4$, $123_6$, $AB_{16}$ et $FF_{16}$

101001_2=41, 

110100_2=52, 

123_4=27, 

1230_4=108,

1010_4=68,

123_6=51



### 1.2 Méthode de gauche à droite (méthode de Horner)

Voici un 2ème algorihtme permettant aussi de convertir un nombre entier écrit en base $b$ vers la base décimale mais en parcourant le nombre dans l'autre sens :
> resultat $\leftarrow$ 0<br>
> **Pour** tous les chiffres $c$ allant de gauche à droite : <br>
>>    resultat $\leftarrow$ b $\times$ resultat<br>
>>    resultat $\leftarrow$ resultat + $c$<br>
>
> Renvoyer resultat<br>

**Remarque :** Cette méthode peut paraître plus complexe que la première mais elle est en fait bien plus efficace (rapide) car elle ne nécessite que des opérations simples (multiplication et addition) au lieu d'utiliser des puissances.

**Q3. Programmer une fonction ``convert_horner`` qui, selon l'algorithme précédent, prend en argument un nombre entier et la base dans laquelle il est écrit et renvoie sa valeur décimale. Penser à spécifier la fonction et éventuellement à utiliser ``assert``.**

In [77]:
# Question 3
        #resultat=resultat*b
        #resultat=resultat+c
        
        #resultat=resultat*b
        #resultat=resultat+int(c[l-1])

c=(input("Entrer un nombre dans la base d'origine: "))
b=int(input("Donner la base de votre nombre: "))
    
def convert_horner(c,b):
    """Cette fonction permet de convertir un nombre d'une base spécifiée à la base décimale (pas hexadecimal)."""
    assert b>=2, "Base doit être supérieure ou égale à 2"
    
    resultat=0
    i=0
    l=len(c)
    
    for i in range(0,l,1):
        resultat=resultat+int(c[l-1-i])*b**i
        print (resultat, i,l)
    return resultat
    
result=convert_horner(c,b)
print(result)

Entrer un nombre dans la base d'origine: 111111110
Donner la base de votre nombre: 2
0 0 9
2 1 9
6 2 9
14 3 9
30 4 9
62 5 9
126 6 9
254 7 9
510 8 9
510


**Q4. Utiliser la fonction``convert_horner`` précédente pour donner la valeur décimale des nombres suivants :**
$1000\,0011_2$, $1010\,0100_2$, $1111\,1111_2$ et $1\,1111\,1110_{2}$

10000011_2=131,
10100100_2=164,
11111111_2=255,
111111110_2=510,

## 2. Conversion la base décimale vers une base b

### 2.1 Conversion par divisions successives par $b$
Pour convertir un nombre entier de la base décimale vers une base $b$, il suffit de le diviser par $b$ (division entière), de noter le reste puis de diviser à nouveau le résultat par $b$ et de noter le reste et ainsi de suite jusqu'à obtenir 0. En écrivant les restes obtenus en commençant par le dernier, on obtient le résultat de la conversion.

**Q5. Programmer une fonction ``convert_base`` qui, selon l'algorithme précédent, prend en argument un nombre entier dans la base décimale et la base dans laquelle on veut le convertir et renvoie la valeur convertie. Penser à spécifier la fonction et éventuellement à utiliser ``assert``.**

In [6]:
# Question 5

c=int(input("Entrer un nombre dans la base 10: "))
b=int(input("Entrer la base vers laquelle 'c' sera converti: "))

def convert_base(c,b):
    resultat=0
    i=0
    l=str(c)
    
    for i in range():
        c%b

**Q6. Utiliser la fonction``convert_base`` précédente pour convertir le nombre 213 dans les bases 2, 4 et 16.**

In [7]:
# Question 6

### 2.2 Conversion par divisions successives par $b^i$
On utilisera plutôt cette méthode si on veut fixer le nombre $k$ de chiffres utilisés pour écrire le résultat, quitte à ce que ce résultat commence par des 0.

Il suffit alors d'effectuer la division entière du nombre à convertir par $b^{k-1}$ puis de recommencer en divisant le reste par $b^{k-2}$ et ainsi de suite judsqu'à diviser par $b^0=1$. Les résultats successifs de ces divisions dans l'ordre forment le nombre final. Evidemment, cela ne fonctionne pas si le nombre à convertir est trop grand pour le nombre de chiffres choisi pour le résultat.

Dans le cas d'une conversion vers le binaire, l'agorithme est le suivant :
> **Pour** $i$ allant de $k-1$ à $0$ : <br>
>> **Si** $n \geq b^i$ :<br>
>>>    $n \leftarrow n-b^i $<br>
>>>    Rajouter le chiffre $1$ au résultat<br>
>>
>> **Sinon** :<br>
>>> Rajouter le chiffre $0$ au résultat<br>
>
> Renvoyer resultat<br>

**Q7. Programmer une fonction ``convert_bin`` qui, selon l'algorithme précédent, prend en argument un nombre entier dans la base décimale et le nombre de chiffres souhaité pour le résultat et renvoie la valeur convertie en binaire. Penser à spécifier la fonction et éventuellement à utiliser ``assert``.**

In [8]:
# Question 7

**Q8. Utiliser la fonction``convert_bin`` précédente pour convertir les nombres 197, 63 et 78 en nombres binaires de 8 chiffres.**

In [9]:
# Question 8